File: src/test/java/hzt/aoc/day17/Part1ConwayCubesTest.java
package hzt.aoc.day17;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

class Part1ConwayCubesTest {

    private final Part1ConwayCubes conwayCubes = new Part1ConwayCubes();

    @Test
    void countActiveNeighbors() {
        final int[][] input = {
                {0, 1, 0},
                {0, 0, 1},
                {1, 1, 1},
        };
        final List<List<List<Boolean>>> grid = convertIntArrayToBooleanList(input);
        System.out.println(conwayCubes.grid3DAsString(grid));
        final int neighbours = conwayCubes.countActiveNeighbors(new Point3D(1, 2, 0), grid);
        Assertions.assertEquals(3, neighbours);
    }

    List<List<List<Boolean>>> convertIntArrayToBooleanList(final int[][] input) {
        final List<List<List<Boolean>>> grid3D = new ArrayList<>();
        final List<List<Boolean>> grid2D = new ArrayList<>();
        for (final int[] intRow : input) {
            final List<Boolean> row = new ArrayList<>();
            for (final int value : intRow) {
                row.add(value == 1);
            }
            grid2D.add(row);
        }
        grid3D.add(grid2D);
        return grid3D;
    }
}

File: src/test/kotlin/hzt/aoc/day17/Part1ConwayCubesTest.kt
package hzt.aoc.day17

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class Part1ConwayCubesTest {
    private val conwayCubes = Part1ConwayCubes()
    @Test
    fun countActiveNeighbors() {
        val input = arrayOf(intArrayOf(0, 1, 0), intArrayOf(0, 0, 1), intArrayOf(1, 1, 1))
        val grid = convertIntArrayToBooleanList(input)
        println(conwayCubes.grid3DAsString(grid))
        val neighbours = conwayCubes.countActiveNeighbors(Point(1, 2, 0), grid)
        Assertions.assertEquals(3, neighbours)
    }

    private fun convertIntArrayToBooleanList(input: Array<IntArray>):
            MutableList<MutableList<MutableList<Boolean>>> {
        val grid3D: MutableList<MutableList<MutableList<Boolean>>> = ArrayList()
        val grid2D: MutableList<MutableList<Boolean>> = ArrayList()
        for (intRow in input) {
            val row: MutableList<Boolean> = ArrayList()
            for (value in intRow) {
                row.add(value == 1)
            }
            grid2D.add(row)
        }
        grid3D.add(grid2D)
        return grid3D
    }
}
